### PR TITLE
plumbing: format/packfile, fix infinite recursion in fixAndBreakChainsOne

### PR DESCRIPTION
**Pull request by lukemarsden**
## Summary

The `fixAndBreakChainsOne` function could enter infinite recursion when processing cyclic delta chains in packfiles. This happened because the cycle guard (checking `otp.Base != nil`) was ineffective - `Base` is only set AFTER the recursive call completes.

If a packfile contained a cyclic delta chain (A→B→A), the function would recurse infinitely:

1. Process A, recurse to process base B
2. Process B, recurse to process base A  
3. A.Base is still nil (not set yet), so recurse again
4. Stack overflow (runtime: goroutine stack exceeds 1000000000-byte limit)

## Fix

Add a `visiting` map that tracks objects currently in the recursion path. If we encounter an object we're already visiting, we've found a cycle and break it by calling `undeltify()` on the object.

## Testing

Existing tests pass. The fix handles an edge case that shouldn't occur in well-formed packfiles, but prevents catastrophic failure when it does.

### CODE REVIEW

**Comment by lukemarsden** [comment]
can anyone advise if this fix is correct please?

File: plumbing/format/packfile/delta_selector.go [hunk 1]
```diff
@@ -138,16 +138,19 @@ func (dw *deltaSelector) fixAndBreakChains(objectsToPack []*ObjectToPack) error
 		m[otp.Hash()] = otp
 	}
 
+	// visiting tracks objects currently in the recursion path to detect cycles
+	visiting := make(map[plumbing.Hash]bool)
```
Comment (pjbgf) [inline plumbing/format/packfile/delta_selector.go:142]:
```suggestion
	// track objects to break away from circular references.
	seen := make(map[plumbing.Hash]struct{})
```

File: plumbing/format/packfile/delta_selector.go [hunk 2]
```diff
@@ -138,16 +138,19 @@ func (dw *deltaSelector) fixAndBreakChains(objectsToPack []*ObjectToPack) error
...
+
 	for _, otp := range objectsToPack {
-		if err := dw.fixAndBreakChainsOne(m, otp); err != nil {
+		if err := dw.fixAndBreakChainsOne(m, otp, visiting); err != nil {
```
Comment (pjbgf) [inline plumbing/format/packfile/delta_selector.go:145]:
```suggestion
		if err := dw.fixAndBreakChainsOne(m, otp, seen); err != nil {
```

File: plumbing/format/packfile/delta_selector.go [hunk 3]
```diff
@@ -138,16 +138,19 @@ func (dw *deltaSelector) fixAndBreakChains(objectsToPack []*ObjectToPack) error
...
 			return err
 		}
 	}
 
 	return nil
 }
 
-func (dw *deltaSelector) fixAndBreakChainsOne(objectsToPack map[plumbing.Hash]*ObjectToPack, otp *ObjectToPack) error {
+func (dw *deltaSelector) fixAndBreakChainsOne(objectsToPack map[plumbing.Hash]*ObjectToPack, otp *ObjectToPack, visiting map[plumbing.Hash]bool) error {
```
Comment (pjbgf) [inline plumbing/format/packfile/delta_selector.go:153]:
```suggestion
func (dw *deltaSelector) fixAndBreakChainsOne(objectsToPack map[plumbing.Hash]*ObjectToPack, otp *ObjectToPack, seen map[plumbing.Hash]struct{}) error {
```

File: plumbing/format/packfile/delta_selector.go [hunk 4]
```diff
@@ -159,6 +162,14 @@ func (dw *deltaSelector) fixAndBreakChainsOne(objectsToPack map[plumbing.Hash]*O
 		return nil
 	}
 
+	hash := otp.Hash()
+
+	// Cycle detection: if we're already visiting this object in the current
+	// recursion path, we have a cyclic delta chain. Break it by undeltifying.
+	if visiting[hash] {
```
Comment (pjbgf) [inline plumbing/format/packfile/delta_selector.go:169]:
```suggestion
	if _, ok := seen[hash]; ok {
```

File: plumbing/format/packfile/delta_selector.go [hunk 5]
```diff
@@ -173,10 +184,16 @@ func (dw *deltaSelector) fixAndBreakChainsOne(objectsToPack map[plumbing.Hash]*O
 		return dw.undeltify(otp)
 	}
 
-	if err := dw.fixAndBreakChainsOne(objectsToPack, base); err != nil {
+	// Mark this object as being visited before recursing
+	visiting[hash] = true
```
Comment (pjbgf) [inline plumbing/format/packfile/delta_selector.go:188]:
```suggestion
	seen[hash] = struct{}{}
```

File: plumbing/format/packfile/delta_selector.go [hunk 6]
```diff
@@ -173,10 +184,16 @@ func (dw *deltaSelector) fixAndBreakChainsOne(objectsToPack map[plumbing.Hash]*O
...
+
+	if err := dw.fixAndBreakChainsOne(objectsToPack, base, visiting); err != nil {
 		return err
 	}
 
+	// Done visiting this object
+	delete(visiting, hash)
```
Comment (pjbgf) [inline plumbing/format/packfile/delta_selector.go:195]:
```suggestion
	delete(seen, hash)
```

**Review by pjbgf — CHANGES_REQUESTED** [review]
@lukemarsden thanks for looking into this. I tested your changes and it does what it says on the tin.

However we may need to make a few changes (as per below) to align with the codebase at large.

Please add a test to confirm the expected behaviour and to avoid future regression. For example:
```golang
type mockDeltaObject struct {
	plumbing.EncodedObject
	baseHash   plumbing.Hash
	actualHash plumbing.Hash
	actualSize int64
}

func (m *mockDeltaObject) BaseHash() plumbing.Hash {
	return m.baseHash
}

func (m *mockDeltaObject) ActualHash() plumbing.Hash {
	return m.actualHash
}

func (m *mockDeltaObject) ActualSize() int64 {
	return m.actualSize
}

func (m *mockDeltaObject) Type() plumbing.ObjectType {
	return plumbing.REFDeltaObject
}

func (s *DeltaSelectorSuite) TestObjectsToPackCircularReference() {
	objA := newObject(plumbing.BlobObject, []byte("object A content"))
	hashA := objA.Hash()

	objB := newObject(plumbing.BlobObject, []byte("object B content"))
	hashB := objB.Hash()

	_, err := s.store.SetEncodedObject(objA)
	s.Require().NoError(err)
	_, err = s.store.SetEncodedObject(objB)
	s.Require().NoError(err)

	// Create mock delta objects with circular references and override
	// their values in the store.
	deltaA := &mockDeltaObject{
		EncodedObject: objA,
		baseHash:      hashB,
		actualHash:    hashA,
		actualSize:    objA.Size(),
	}
	s.store.Objects[hashA] = deltaA

	deltaB := &mockDeltaObject{
		EncodedObject: objB,
		baseHash:      hashA,
		actualHash:    hashB,
		actualSize:    objB.Size(),
	}
	s.store.Objects[hashB] = deltaB

	otp, err := s.ds.ObjectsToPack([]plumbing.Hash{hashA, hashB}, 10)
	s.Require().NoError(err)
	s.Len(otp, 2)
}

```

**Comment by pjbgf** [comment]
Closing as https://github.com/go-git/go-git/issues/2249 supersedes this PR.
